src/lib/mongoBackend/MongoCommonUpdate.cpp

```diff
@@ -629,28 +506,44 @@ static bool appendAttribute(BSONObj& attrs, BSONObj& newAttrs, ContextAttribute*
 * name
 *
 */
-static bool legalIdUsage(BSONObj& attrs, ContextAttribute* caP) {
+static bool legalIdUsage(BSONObj& attrs, ContextAttribute* caP)
+{
 
-    if (caP->getId() == "") {
-        /* Attribute attempting to append hasn't ID. Thus, no attribute with same name can have ID in attrs */
-        for( BSONObj::iterator i = attrs.begin(); i.more(); ) {
-            BSONObj attr = i.next().embeddedObject();
-            if (STR_FIELD(attr, ENT_ATTRS_NAME) == caP->name && STR_FIELD(attr, ENT_ATTRS_ID) != "") {
-                return false;
-            }
-        }
-        return true;
+  std::string prefix = caP->name + "__";
+  if (caP->getId() == "")
+  {
+    /* Attribute attempting to append hasn't ID. Thus, no attribute with same name can have ID in attrs,
+     * i.e. no attribute starting with "<attrName>" can at the same time start with "<attrName>__" */
+    std::set<std::string> attrNames;
+    attrs.getFieldNames(attrNames);
+    for( std::set<std::string>::iterator i = attrNames.begin(); i!=attrNames.end(); ++i)
+    {
+      std::string attrName = *i;
+      if (strncmp(caP->name.c_str(), attrName.c_str(), strlen(caP->name.c_str())) == 0 &&
+          strncmp(prefix.c_str(), attrName.c_str(), strlen(prefix.c_str())) == 0)
+      {
+        return false;
+      }
     }
-    else {
-        /* Attribute attempting to append has ID. Thus, no attribute with same name cannot have ID in attrs */
-        for( BSONObj::iterator i = attrs.begin(); i.more(); ) {
-            BSONObj attr = i.next().embeddedObject();
-            if (STR_FIELD(attr, ENT_ATTRS_NAME) == caP->name && STR_FIELD(attr, ENT_ATTRS_ID) == "") {
-                return false;
-            }
-        }
-        return true;
+    return true;
+  }
+  else
+  {
+    /* Attribute attempting to append has ID. Thus, no attribute with same name cannot have ID in attrs,
+     * i.e. no attribute starting with "<attrName>" can at the same time have a name not starting with "<attrName>__" */
+    std::set<std::string> attrNames;
+    attrs.getFieldNames(attrNames);
+    for( std::set<std::string>::iterator i = attrNames.begin(); i!=attrNames.end(); ++i)
+    {
+      std::string attrName = *i;
+      if (strncmp(caP->name.c_str(), attrName.c_str(), strlen(caP->name.c_str())) == 0 &&
```

Comment:
strlen(caP->name.c_str())   == caP->name.length()

We are (especially myself) very bad at using the methods of std::string.
We almost always 'go back' to the known char* and the standard C library.
Better to learn more about std::string, now that we are using it ... :-)